### DISCUSSION TITLE
Unable to install Linux headers in GitHub action

### DISCUSSION THREAD

**Discussion by InterLinked1** [community/community#28607] [Actions]
I've never been able to get the kernel headers to install inside of GitHub Actions. I've tried before and always had to revert because the headers can't be installed. Specifically, installing the kernel headers using the canonical command:

```
apt-get install -y linux-headers-`uname -r`
```

Currently, I have a GitHub Action that does software builds for userland software, but it has an optional dependency on kernel software which currently can't be built inside a GitHub Action because the headers do not install using the above command. This is disappointing because the kernel side of things tends to be more fickle than the userland stuff which is much more regularly tested, so it's difficult to catch bugs and compilation issues with the kernel software.

```
E: Unable to locate package linux-headers-5.15.0-1014-azure
E: Couldn't find any package by glob 'linux-headers-5.15.0-1014-azure'
E: Couldn't find any package by regex 'linux-headers-5.15.0-1014-azure'
```

I am using a Debian 11 container since Debian isn't a native runner platform (plus full system access is required for the build). Here's an example repo demonstrating this issue: https://github.com/InterLinked1/debian-headers-ci

I did some testing and similar packages to the above *are* available:
```
Your kernel: 5.15.0-1014-azure
Available headers for your system:
linux-headers-5.10.0-12-amd64/stable-security 5.10.103-1 amd64
linux-headers-5.10.0-12-cloud-amd64/stable-security 5.10.103-1 amd64
linux-headers-5.10.0-12-common/stable-security 5.10.103-1 all
linux-headers-5.10.0-12-common-rt/stable-security 5.10.103-1 all
linux-headers-5.10.0-12-rt-amd64/stable-security 5.10.103-1 amd64
linux-headers-5.10.0-13-amd64/stable 5.10.106-1 amd64
linux-headers-5.10.0-13-cloud-amd64/stable 5.10.106-1 amd64
linux-headers-5.10.0-13-common/stable 5.10.106-1 all
linux-headers-5.10.0-13-common-rt/stable 5.10.106-1 all
linux-headers-5.10.0-13-rt-amd64/stable 5.10.106-1 amd64
linux-headers-5.10.0-14-amd64/stable-security 5.10.113-1 amd64
linux-headers-5.10.0-14-cloud-amd64/stable-security 5.10.113-1 amd64
linux-headers-5.10.0-14-common/stable-security 5.10.113-1 all
linux-headers-5.10.0-14-common-rt/stable-security 5.10.113-1 all
linux-headers-5.10.0-14-rt-amd64/stable-security 5.10.113-1 amd64
linux-headers-5.10.0-15-amd64/stable-security 5.10.120-1 amd64
linux-headers-5.10.0-15-cloud-amd64/stable-security 5.10.120-1 amd64
linux-headers-5.10.0-15-common/stable-security 5.10.120-1 all
linux-headers-5.10.0-15-common-rt/stable-security 5.10.120-1 all
linux-headers-5.10.0-15-rt-amd64/stable-security 5.10.120-1 amd64
linux-headers-5.10.0-16-amd64/stable-security 5.10.127-2 amd64
linux-headers-5.10.0-16-cloud-amd64/stable-security 5.10.127-2 amd64
linux-headers-5.10.0-16-common/stable-security 5.10.127-2 all
linux-headers-5.10.0-16-common-rt/stable-security 5.10.127-2 all
linux-headers-5.10.0-16-rt-amd64/stable-security 5.10.127-2 amd64
linux-headers-amd64/stable-security 5.10.127-2 amd64
linux-headers-cloud-amd64/stable-security 5.10.127-2 amd64
linux-headers-rt-amd64/stable-security 5.10.127-2 amd64
```

However, the actual OS has no native package available for it.

It'd be really great to be able to set up CI for the kernel code. I'm wondering:

- Is this a bug with the Azure container images that GitHub is using? On real hardware (or virtual machines running on-prem), we've never had these issues. No issue with DigitalOcean droplets either.
- Is there any way this can be fixed, or any workaround?

Thanks!

**Comment by airtower-luna** [comment]
The assumption that
```
apt-get install -y linux-headers-`uname -r`
```
is the "canonical" command to install kernel headers is incorrect. That command tries to install the headers *for the exact running kernel version*, which is what causes your problem. Because Debian doesn't have whatever Ubuntu-based and patched kernel version Azure is running.

> but it has an optional dependency on kernel software which currently can't be built inside a GitHub Action because the headers do not install using the above command.

If by "kernel software" you mean a kernel module, that will indeed be a problem. In that case you should forget about using a job container, messing with the kernel from a container is a bad idea in general. If you build directly in the runner VM I'd hope whoever patched the kernel also had the sense to provide a matching headers package.

Otherwise, if your software just uses kernel interfaces and needs to include the headers defining those: The common approach is to install the appropriate kernel headers package for your platform, in this case `linux-headers-amd64`. That'll pull in the latest version in the Debian repository, and (important in non-container situations) it'll get updated like any other package. In the vast majority of cases the version difference doesn't matter because "don't break userspace" is one of the key principles of Linux kernel development.

I'm really surprised you haven't had those issues before either way, are all the other machines on which you've been running similar container workflows also running Debian 11 or some derivate?

**Reply by InterLinked1** [reply]
> is the "canonical" command to install kernel headers is incorrect. That command tries to install the headers for the exact running kernel version, which is what causes your problem. Because Debian doesn't have whatever Ubuntu-based and patched kernel version Azure is running.

Well, I was wondering about that. This is how I've always seen it done, however, and it seems it's quite often that they will assume this exact dependency, meaning that if that fails, the software will not build at all, at least not without serious monkeying around with the way the Makefiles are written, and possibly who knows what else.

> but it has an optional dependency on kernel software which currently can't be built inside a GitHub Action because the headers do not install using the above command.

> If by "kernel software" you mean a kernel module, that will indeed be a problem. In that case you should forget about using a job container, messing with the kernel from a container is a bad idea in general. If you build directly in the runner VM I'd hope whoever patched the kernel also had the sense to provide a matching headers package.

I guess I could try that - I picked a container since the primary development and actually used platform is Debian. Ubuntu might be hopefully similar enough that that might work, but we've also recently found behavior differences between Debian and Ubuntu so being able to test on Debian would be ideal.

> Otherwise, if your software just uses kernel interfaces and needs to include the headers defining those: The common approach is to install the appropriate kernel headers package for your platform, in this case linux-headers-amd64. That'll pull in the latest version in the Debian repository, and (important in non-container situations) it'll get updated like any other package. In the vast majority of cases the version difference doesn't matter because "don't break userspace" is one of the key principles of Linux kernel development.

Unfortunately, that's not the case - I can install different headers manually by seeing what's available and then picking one that's close, but as soon as you start compiling the software, it whines that it can't find the *exact* kernel headers it's looking for and bails out.

> I'm really surprised you haven't had those issues before either way, are all the other machines on which you've been running similar container workflows also running Debian 11 or some derivate?

I don't really use containers outside of GitHub Actions, just virtual machines for the most part. And no, I've generally not had an issue with the `uname -a` kernel headers not being found on Debian 8, 9, 10, 11, or many other Linux distros (possibly even BSDs, don't remember off hand).

There was a brief period of time where I could not get the kernel headers to install on a Debian 10 VPS in DigitalOcean, but a few months later the package must've been added as I was able to install them at a later point. In this case, it seems more like a case of not being added at all, hence my wondering if I need to explicitly target a GitHub Action platform that can install the native kernel headers.

**Reply by airtower-luna** [reply]
> I don't really use containers outside of GitHub Actions, just virtual machines for the most part.

Yeah, in that case installing the matching headers will just work as long as you're using the distro kernel, because of course the distro provides matching headers. In a container the *running* kernel is usually the *only* thing shared with the host – except by chance, or intentionally deploying containers to matching hosts (which kind of breaks the concept).
